crypto/hd: DerivePrivateKeyForPath can panic with index out of bounds if the path ends with a slash "/", due to improper check of length

<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺ 
v                            ✰  Thanks for opening an issue! ✰    
v    Before smashing the submit button please review the template.
v    Please also ensure that this is not a duplicate issue :)  
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

<!--
IMPORTANT: Prior to opening a bug report, check if it affects one of the core modules
and if its elegible for a bug bounty on `SECURITY.md`. Bugs that are not submitted
through the appropriate channels won't receive any bounty.
 -->

## Summary of Bug
Found by auditing the code for Stargate and preparing to fuzz it, if any path is provided with a trailing slash for example just by simply passing in "m/16/19/" instead of "m/16/19" or even "/", or "m/16//10"
```go
func TestDerivePrivateKeyForPath(t *testing.T) {
        hd.DerivePrivateKeyForPath([32]byte{}, [32]byte{}, "m/5/")
}
```

results in this panic
```shell
--- FAIL: TestDerivePrivateKeyForPath (0.00s)
panic: runtime error: slice bounds out of range [-1:] [recovered]
	panic: runtime error: slice bounds out of range [-1:]

goroutine 36 [running]:
testing.tRunner.func1.2(0x18b9080, 0xc00024bea8)
	/Users/emmanuelodeke/go/src/go.googlesource.com/go/src/testing/testing.go:1144 +0x332
testing.tRunner.func1(0xc000103080)
	/Users/emmanuelodeke/go/src/go.googlesource.com/go/src/testing/testing.go:1147 +0x4b6
panic(0x18b9080, 0xc00024bea8)
	/Users/emmanuelodeke/go/src/go.googlesource.com/go/src/runtime/panic.go:965 +0x1b9
github.com/cosmos/cosmos-sdk/crypto/hd.DerivePrivateKeyForPath(0x0, 0x0, 0x0, 0x0, 0xed2c7f8abb74f3c9, 0x8f0215d93407b907, 0x7a28ae7c374e8027, 0xb869ce287cabd893, 0x191cada, 0x4, ...)
	/Users/emmanuelodeke/go/src/github.com/cosmos/cosmos-sdk/crypto/hd/hdpath.go:192 +0x587
github.com/cosmos/cosmos-sdk/crypto/hd_test.TestDerivePrivateKeyForPath(0xc000103080)
	/Users/emmanuelodeke/go/src/github.com/cosmos/cosmos-sdk/crypto/hd/hdpath_test.go:286 +0x4d
testing.tRunner(0xc000103080, 0x1a80808)
	/Users/emmanuelodeke/go/src/go.googlesource.com/go/src/testing/testing.go:1194 +0xef
created by testing.(*T).Run
	/Users/emmanuelodeke/go/src/go.googlesource.com/go/src/testing/testing.go:1239 +0x2b3
exit status 2
FAIL	github.com/cosmos/cosmos-sdk/crypto/hd	0.296s
```

## Version
All versions with this code
<!-- git commit hash or release version -->

## Cause
This code https://github.com/cosmos/cosmos-sdk/blob/c54025dac01ba36224b0eaff1a0030afe939b22c/crypto/hd/hdpath.go#L190-L192
assumes that ALL the parts will have a non-empty string, but that's clearly not true, if a path with a trailing slash were passed it, it'd crash.

## Remedy
Before that `part[len(part)-1] == "'"`, we need to check that the segment is non-empty. There is 2 fold remedy:
* replace "//" with "/"
* strip trailing and starting slashes

Let's please backport this change to all releases like Stargate et al.

## For Admin Use

- [x] Not duplicate issue
- [x] Appropriate labels applied
- [x] Appropriate contributors tagged
- [ ] Contributor assigned/self-assigned

--- a/crypto/hd/hdpath.go
+++ b/crypto/hd/hdpath.go
@@ -6,6 +6,7 @@ import (
 	"encoding/binary"
 	"fmt"
 	"math/big"
+	"path/filepath"
 	"strconv"
 	"strings"
 
@@ -177,6 +178,9 @@ func ComputeMastersFromSeed(seed []byte) (secret [32]byte, chainCode [32]byte) {
 // DerivePrivateKeyForPath derives the private key by following the BIP 32/44 path from privKeyBytes,
 // using the given chainCode.
 func DerivePrivateKeyForPath(privKeyBytes, chainCode [32]byte, path string) ([]byte, error) {
+	// First step is to trim the right end path separator lest we panic.
+	// See issue https://github.com/cosmos/cosmos-sdk/issues/8557
+	path = strings.TrimRightFunc(path, func(r rune) bool { return r == filepath.Separator })
 	data := privKeyBytes
 	parts := strings.Split(path, "/")
 
@@ -187,7 +191,10 @@ func DerivePrivateKeyForPath(privKeyBytes, chainCode [32]byte, path string) ([]b
 		parts = parts[1:]
 	}
 
-	for _, part := range parts {
+	for i, part := range parts {
+		if part == "" {
+			return nil, fmt.Errorf("path %q with split element #%d is an empty string", part, i)
+		}
 		// do we have an apostrophe?
 		harden := part[len(part)-1:] == "'"
 		// harden == private derivation, else public derivation:

--- a/crypto/hd/hdpath_test.go
+++ b/crypto/hd/hdpath_test.go
@@ -281,3 +281,26 @@ func ExampleSomeBIP32TestVecs() {
 	//
 	// c4c11d8c03625515905d7e89d25dfc66126fbc629ecca6db489a1a72fc4bda78
 }
+
+// Ensuring that we don't crash if values have trailing slashes
+// See issue https://github.com/cosmos/cosmos-sdk/issues/8557.
+func TestDerivePrivateKeyForPathDoNotCrash(t *testing.T) {
+	paths := []string{
+		"m/5/",
+		"m/5",
+		"/44",
+		"m//5",
+		"m/0/7",
+		"/",
+		" m       /0/7",          // Test case from fuzzer
+		"              /       ", // Test case from fuzzer
+		"m///7//////",
+	}
+
+	for _, path := range paths {
+		path := path
+		t.Run(path, func(t *testing.T) {
+			hd.DerivePrivateKeyForPath([32]byte{}, [32]byte{}, path)
+		})
+	}
+}

